### PR TITLE
[HttpFoundation] add `HeaderUtils::parseQuery()`: it does the same as `parse_str()` but preserves dots in variable names

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

Inspired by https://github.com/symfony/psr-http-message-bridge/pull/80
/cc @drupol 

Related to #9009, #29664, #26220 but also https://github.com/api-platform/core/issues/509 and https://www.drupal.org/project/drupal/issues/2984272
/cc @dunglas @alexpott

### CODE REVIEW

**Comment by drupol** [comment]
Ho nice ! 

We should add a test with spaces in the key parameter.

File: src/Symfony/Component/HttpFoundation/Tests/HeaderUtilsTest.php [hunk 1]
```diff
@@ -129,4 +129,27 @@ public function provideMakeDispositionFail()
...
+    {
+        $this->assertSame($expected ?? $queryString, http_build_query(HeaderUtils::parseQueryString($queryString), '', '&'));
+    }
+
+    public function provideParseQueryString()
+    {
+        return [
+            ['a=b&c=d'],
+            ['a.b=c'],
+            ['a+b=c'],
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Tests/HeaderUtilsTest.php:146]: @drupol this is the test case with a space in the name

**Review by drupol — APPROVED** [review]
I'm glad I inspired you for this feature, I couldn't agree more with it.

It will help us at [European Commission](https://github.com/ecphp) because we are using [API Platform](https://api-platform.com/) and [CAS authentication](https://github.com/ecphp/cas-bundle).
Both are relying on query parameters and they cannot be changed during the request, this is reason why a middleware [symfony/psr-http-message-bridge](https://github.com/symfony/psr-http-message-bridge) was used, so we could override it with [loophp/unaltered-psr-http-message-bridge-bundle](https://github.com/loophp/unaltered-psr-http-message-bridge-bundle).

Once this feature will be merged, I guess it will help a bunch of people and remove some hacks here and there.

However, I really hope this behavior will be updated in PHP 8, I couldn't believe it was working as such when I discovered this issue. It's crazy!

I also tested the PR here: https://3v4l.org/VRCoh

The example compare the behavior of [\parse_str()](https://www.php.net/manual/en/function.parse-str.php) and the custom `HeaderUtils::queryParseString()` you're introducing.

I did this example because I wanted to check why you were testing the `a\0b` binary string, I guess this would unlikely happens in real world right ?

I'm also wondering if it wouldn't be easier to parse and convert in every key parameters with [\bin2hex()](https://www.php.net/manual/en/function.bin2hex.php), and then let [\parse_str()](https://www.php.net/manual/en/function.parse-str.php) do its job, and then convert back. It would be also maybe faster than going to all those `if` conditions ? WDYT? 

Also, do you think such a feature will be backported in Symfony < 5.2 ?

File: src/Symfony/Component/HttpFoundation/HeaderUtils.php [hunk 2]
```diff
@@ -193,6 +193,54 @@ public static function makeDisposition(string $disposition, string $filename, st
...
+            if (false === $i = strpos($k, '[')) {
+                $k = bin2hex($k);
+            } else {
+                $k = substr_replace($k, bin2hex(substr($k, 0, $i)), 0, $i);
+            }
+
+            $query[] = $k.$v;
+        }
+
+        parse_str(implode('&', $query), $query);
```
Comment (drupol) [inline src/Symfony/Component/HttpFoundation/HeaderUtils.php:229]: How about using this ?

```suggestion
        parse_str(http_build_query($query), $query);
```

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpFoundation/HeaderUtils.php:229]: Not possible as this would lead to double URL encoding.

**Comment by nicolas-grekas** [comment]
> the a\0b binary string, I guess this would unlikely happens in real world right ?

Yes, that's a very unlikely edge case of parse_str() that I duplicated for parity.

> parse and convert in every key parameters with \bin2hex(), and then let \parse_str() do its job, and then convert back. It would be also maybe faster than going to all those if conditions ? WDYT?

I'm not sure what you mean sorry. About perf, those if are going to be very fast, much faster than using a regexp. But please prove me wrong if you want to give your idea a try!

> Also, do you think such a feature will be backported in Symfony < 5.2 ?

I don't think it should, that's a new feature.

One should note also that I didn't change the createFromGlobals method. This means that dots will still be replaced by default. This is important to preserve BC. But at least ppl that care now have a helper to opt-in for the fixed parser.

**Comment by gmponos** [comment]
Would you be interested into making this function work also with a query string like this? 

https://3v4l.org/86Tu5 and return an array containing `'test' => ['what', 'what2']`

Since this is a valid URL http://localhost/myendpoint?test=what&test=what2 

related also with my comment [here](https://github.com/symfony/symfony/issues/37029#issuecomment-636520304)

**Comment by nicolas-grekas** [comment]
@gmponos I've been wondering about this also. This would be a too big API change when dealing with query/cookie/etc bags. But I whish we'll find a way to get there eventually yes (getting rid of `[]` to access to multiple submitted values.)

**Comment by dunglas** [comment]
> But I whish we'll find a way to get there eventually yes

The Go API is nice to handle these cases: https://golang.org/pkg/net/url/#URL.Query

Maybe could this set of features be included in the proposed URI component (#36999)?

**Comment by drupol** [comment]
> Would you be interested into making this function work also with a query string like this?
> 
> https://3v4l.org/86Tu5 and return an array containing `'test' => ['what', 'what2']`
> 
> Since this is a valid URL http://localhost/myendpoint?test=what&test=what2
> 
> Sorry if it already does that and I missed it.
> 
> related also with my comment [here](https://github.com/symfony/symfony/issues/37029#issuecomment-636520304)

I'm about to make a regex that does it all, I will submit my snippet tonight probably.

This is the Father's day today and I might be away from computer...

**Comment by nicolas-grekas** [comment]
> At a glance, we should always return an array of arrays:

I'm not sure how this should play with the `[]` syntax. Eg how should `a[b]=c&a[b]=d` be parsed? There are several possible answers. This means this is not for this PR at least :) votes welcome.

**Comment by dunglas** [comment]
The `[]` syntax is a PHP oddity, which isn't covered by any RFC. In "standard mode" (the component), I suggest to **not** support it. We could maybe also propose a PHP mode that just use `parse_str()` and all is oddities (dot replacement, array syntax, etc).

**Comment by nicolas-grekas** [comment]
> The [] syntax is a PHP oddity, which isn't covered by any RFC

For sure - also no RFC should cover what happens server-side.

But we cannot migrate away from it without a serious FC/BC plan.
E.g. the form component relies on this syntax.

**Comment by dunglas** [comment]
The PHP mode could be the default (so it's 100% BC). The "standard-compliant" mode is useful only for a small subset of use cases after all (but for instance both Mercure and Vulcain, as well a many other I-D and RFCs use repeated query parameters without the `[]` prefix, this currently require custom to parse such URLs with PHP/Symfony).

**Comment by nicolas-grekas** [comment]
Would this make sense to you? Then it would be up to users to use this function when they create the request object:

```diff
--- a/src/Symfony/Component/HttpFoundation/HeaderUtils.php
+++ b/src/Symfony/Component/HttpFoundation/HeaderUtils.php
@@ -196,7 +196,7 @@ class HeaderUtils
     /**
      * Like parse_str(), but preserves dots in variable names.
      */
-    public static function parseQuery(string $query, string $separator = '&'): array
+    public static function parseQuery(string $query, bool $phpMode = true, string $separator = '&'): array
     {
         $q = [];
 
@@ -217,6 +217,12 @@ class HeaderUtils
                 $k = substr($k, 0, $i);
             }
 
+            if (!$phpMode) {
+                $q[$k][] = '' === $v ? '' : urldecode(substr($v, 1));
+
+                continue;
+            }
+
             $k = ltrim($k, ' ');
 
             if (false === $i = strpos($k, '[')) {
@@ -226,6 +232,10 @@ class HeaderUtils
             }
         }
 
+        if (!$phpMode) {
+            return $q;
+        }
+
         parse_str(implode('&', $q), $q);
 
         $query = [];
```

**Comment by dunglas** [comment]
LGTM

**Comment by drupol** [comment]
Here I am,

I just made a small example with a `preg_replace_callback()`: https://3v4l.org/d0W4Z

Basically, [the regex](https://regex101.com/r/iXlLWs/1) will convert the relevant part of the keys to hexadecimal, then let parse_str() do its job, then convert back.

There is only one test which produces (`a\0b=c`)a different result, but I guess we will never have this in real life and maybe we can decide all together to skip it if you prefer. Let me know what you think.

**Comment by nicolas-grekas** [comment]
@dunglas updated, please have a look.

@drupol the code in https://3v4l.org/d0W4Z won't parse correctly the [`a%5Bb%5D=c`](https://3v4l.org/1QBUJ) query string + it's going to be way slower.

**Comment by drupol** [comment]
> @drupol the code in https://3v4l.org/d0W4Z won't parse correctly the [`a%5Bb%5D=c`](https://3v4l.org/1QBUJ) query string + it's going to be way slower.

Ok I fixed it... https://3v4l.org/Pam6Z

But ok if it's slower, you decide :-)

**Comment by nicolas-grekas** [comment]
> urldecode($keyValue)

This line will lead to double decoding of the value (once here + twice done by `parse_str()`) - the implem in symfony/psr-http-message-bridge#80 has the same issue btw.

**Comment by drupol** [comment]
> > urldecode($keyValue)
> 
> This line will lead to double decoding of the value (once here + twice done by `parse_str()`) - the implem in [symfony/psr-http-message-bridge#80](https://github.com/symfony/psr-http-message-bridge/pull/80) has the same issue btw.

Well done, I updated it: https://3v4l.org/gMDMV

**Review by drupol — APPROVED** [review]
:+1:

**Comment by simonberger** [comment]
This change heads in a really good direction. Thank you. I have some random comments and questions.
* Does `$ignoreBrackets=true` include the leading whitespace on purpose and why?
* Most likely known but not documented - parseQuery preserves also subsequent whitespace beside dots while parseStr replaces them with `_`  
* What about naming `$ignoreBrackets` more to its achievement? I would have troubles to get an idea of its job without studying the documentation. Some deliberately (too) extreme brainstorming: groupAllParams, alwaysGroup, groupByParameterNames, groupWithoutBrackets alwaysGroupAndIgnoreBrackets.

File: src/Symfony/Component/HttpFoundation/HeaderUtils.php [hunk 3]
```diff
@@ -193,6 +193,64 @@ public static function makeDisposition(string $disposition, string $filename, st
         return $disposition.'; '.self::toString($params, ';');
     }
 
+    /**
+     * Like parse_str(), but preserves dots in variable names.
+     */
+    public static function parseQuery(string $query, bool $ignoreBrackets = false, string $separator = '&'): array
```
Comment (dunglas) [inline src/Symfony/Component/HttpFoundation/HeaderUtils.php:199]: Shouldn't we just use the native `parse_str()` if `$ignoreBracket` is set to `false`? And so rename `
`$ignoreBrackets` to something else such as `$phpCompat = true` or something like that?

Many PHP libraries rely on the fact that dot are replaced by underscores and we may introduce issues by changing this. I would prefer to have a pure PHP mode (with all the oddities, including the dots etc), and a "strict" mode doing nothing more than the `URL` class in JS for instance. It feels wrong to me to have some intermediary modes such as "don't replace the dots but ignore brackets".

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpFoundation/HeaderUtils.php:199]: I'm sorry I'm missing your logic. To me, it feels wrong to parse dots (replacing them with `_`) but ignore brackets.
The purpose of the method is to give you access to the original names and have a toggle to accept multiple values with or without using `[]`.
Libs that expect keys mangled by php should be given arrays mangled by php, ie you wouldn't use this function with them.

**Comment by drupol** [comment]
I noticed that Firefox and Chrome behaves differently when it comes to parse the string `a%00b=c`.

Firefox:
image

Chrome:
image

Shouldn't we mimic that behavior here?

**Comment by nicolas-grekas** [comment]
@drupol there is no use case for the null char in URLs. I'm not even sure it's legal from an RFC pov. php-src uses C-strings internally in `parse_str()`, that's why null chars cut the string. I think we should stick to the behavior of php for this one.

**Comment by nicolas-grekas** [comment]
Now rebased on top of  #37271, ready.

File: src/Symfony/Component/HttpFoundation/Tests/HeaderUtilsTest.php [hunk 4]
```diff
@@ -129,4 +129,41 @@ public function provideMakeDispositionFail()
...
+        ];
+    }
+
+    public function testParseCookie()
+    {
+        $query = 'a.b=c; def%5Bg%5D=h';
+        $this->assertSame($query, http_build_query(HeaderUtils::parseQuery($query, false, ';'), '', '; '));
+    }
+
+    public function testParseueryIgnoreBrackets()
```
Comment (phcorp) [inline src/Symfony/Component/HttpFoundation/Tests/HeaderUtilsTest.php:163]: testParse**Q**ueryIgnoreBrackets

**Comment by nicolas-grekas** [comment]
@simonberger I missed you comment:

> Does $ignoreBrackets=true include the leading whitespace on purpose and why?

fixed, trimming now happens unconditionally.

> Most likely known but not documented - parseQuery preserves also subsequent whitespace beside dots while parseStr replaces them with _

true, but nobody is using spaces in var names so I skipped advertising this.

> What about naming $ignoreBrackets more to its achievement? I would have troubles to get an idea of its job without studying the documentation. Some deliberately (too) extreme name examples: groupAllParams, alwaysGroup, groupByParameterNames, groupWithoutBrackets alwaysGroupAndIgnoreBrackets.

naming... :)
I'm sorry I'm not convinced by any of the current proposals...

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Pull request merged**